### PR TITLE
Use Alarms API

### PR DESCRIPTION
**Pull request by bershanskiy**
Use [Alarms API](https://developer.chrome.com/docs/extensions/reference/alarms/) instead of timeouts because those are not persisted after non-persistent background shutdown. Alarms will survive across background shutdowns and re-activations and will actually re-activate unloaded background.

This patch adds one more permission, `alarms`, which is necessary to use this API (it does not generate any extra permission messages during install). This patch migrates two `setInterval` call sites and leaves [few more](https://github.com/darkreader/darkreader/blob/a5c5c9a6e683419befd87cbadde31525193b589d/src/background/extension.ts#L240) because I plan to refactor those in a different way.

### CODE REVIEW

File: src/background/utils/network.ts [hunk 1]
```diff
@@ -38,12 +38,18 @@ interface CacheRecord {
...
     static TTL = getDuration({minutes: 10});
+    static ALARM_NAME = 'network';
 
     private bytesInUse = 0;
     private records = new Map<string, CacheRecord>();
 
     constructor() {
-        setInterval(() => this.removeExpiredRecords(), getDuration({minutes: 1}));
+        chrome.alarms.onAlarm.addListener(async (alarm) => {
+            if (alarm.name === LimitedCacheStorage.ALARM_NAME) {
```
Comment (Gusted) [inline src/background/utils/network.ts:48]: Hmm, could I ask why we are checking if the the alarm_name is the same? Is this some kind of intentional design of the alarm API?

**Review by Gusted — COMMENTED** [review]
IMO seems like a good change 👍🏽

File: src/background/utils/network.ts (hunk 1 above)
Comment (bershanskiy) [reply src/background/utils/network.ts:48]: Alarm name allows us to react only to the desired alarm, which is helpful when we have multiple different classes creating their own alarms for different things. Without name check every alarm listener would react to every alarm. E.g., `Newsmaker` would react to `LimitedCacheStorage` alarm and query your server for news every minute when you actually want it to do it every 4 hours.

Comment (Gusted) [reply src/background/utils/network.ts:48]: Aha, that makes more change.

**Review by Gusted — APPROVED** [review]
You got my 👍🏽, but because it adds an extra permission, I would like @alexanderby to give an approval for this one as well. As I've seen some developers experiences issues when they request new permissions(even non-powerful-ish ones).

File: src/manifest.json [hunk 2]
```diff
@@ -37,6 +37,7 @@
         }
     ],
     "permissions": [
+        "alarms",
```
Comment (alexanderby) [inline src/manifest.json:40]: One thing I'm afraid it to cause a warning for users after an update.

File: src/utils/time.ts [hunk 3]
```diff
@@ -54,23 +54,27 @@ interface Duration {
...
+        duration += time.hours * 60 * 1000;
     }
     if (time.days) {
-        duration += time.days * 24 * 60 * 60 * 1000;
+        duration += time.days * 24 * 60 * 1000;
     }
     return duration;
 }
 
+export function getDuration(time: Duration) {
```
Comment (alexanderby) [inline src/utils/time.ts:74]: Maybe do it vice versa? Leave the `getDuration` function as is.

File: src/background/utils/network.ts [hunk 4]
```diff
@@ -38,12 +38,18 @@ interface CacheRecord {
...
+                this.removeExpiredRecords();
+            }
+        });
+        chrome.alarms.create(LimitedCacheStorage.ALARM_NAME, {periodInMinutes: getDurationInMinutes({minutes: 1})});
```
Comment (alexanderby) [inline src/background/utils/network.ts:52]: Can you just put `1` here without a function?

**Review by alexanderby — APPROVED** [review]
Nice usage of proper APIs.

File: src/manifest.json (hunk 2 above)
Comment (bershanskiy) [reply src/manifest.json:40]: I don't have time to find all the sources about this, so I'll post them tomorrow. Essentially, `alarms` is one of those odd permisions which must be in the manifest but are granted automatically.

File: src/utils/time.ts [hunk 5]
```diff
@@ -54,25 +54,25 @@ interface Duration {
...
     if (time.days) {
-        duration += time.days * 24 * 60 * 1000;
+        duration += time.days * 24 * 60 * 60 * 1000;
     }
     return duration;
 }
 
-export function getDuration(time: Duration) {
-    return getDurationInMinutes(time) * 60;
+export function getDurationInMinutes(time: Duration) {
```
Comment (alexanderby) [inline src/utils/time.ts:74]: So `getDurationInMinutes` is no more needed, restore the `getDuration` as it was.

**Review by alexanderby — CHANGES_REQUESTED** [review]
Run the `npm run code-style`, it says some variable is not defined.

File: src/utils/time.ts (hunk 5 above)
Comment (alexanderby) [reply src/utils/time.ts:74]: Oops, wrong, it is actually used

File: src/manifest.json (hunk 2 above)
Comment (bershanskiy) [reply src/manifest.json:40]: I have the following evidence that installing update with the new `alarms` permission does not generate extra warnings:
 - `chrome.management.getPermissionWarningsByManifest` creates the list of user-visible permissions warnings for a given manifest and it returns nothing for a manifest with only `alarms` permission ([Chrome docs](https://developer.chrome.com/docs/extensions/reference/management/#method-getPermissionWarningsByManifest), [MDN](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/management/getPermissionWarningsByManifest))
 - Chrome docs do not list `alarms` among ["Permissions with warnings"](https://developer.chrome.com/docs/extensions/mv2/permission_warnings/#permissions_with_warnings)

Ideally, we would be able to test that Chrome and Firefox do not actually just decide to disable extension anyway upon end-user updating the extension. I'm still researching ways to do this.

**Comment by bershanskiy** [comment]
> Run the `npm run code-style`, it says some variable is not defined.

Fixed it. (It said that a wariable was unused because it was a useless import.)

**Review by alexanderby — APPROVED** [review]

**Pull request merged**